devito/ir/equations/algorithms.py

```diff
@@ -147,3 +149,50 @@ def lower_exprs(expressions, **kwargs):
     else:
         assert len(processed) == 1
         return processed.pop()
+
+
+def separate_dimensions(expressions):
+    """
+    Rename Dimensions with clashing names within expressions.
+    """
+    resolutions = {}
+    count = {}  # Keep track of increments on dim names
+    processed = []
+    for e in expressions:
+        # Just want dimensions which appear in the expression
+        # Dimensions in indices
+        dims = set().union(*tuple(set(i.function.dimensions)
+                                  for i in retrieve_indexed(e)))
+        # Dimensions in conditions and ConditionalDimension parents
+        dims = dims.union(*tuple(pull_dims(d.condition, flag=False)
```

Comment:
you probably want to unclutter here and move most of this logic into `pull_dims`

ideally, all you need to do here would be:

```
dims = pull_dims(e, flag=...)

# Sort for groupby
dims = sorted(dims, key=lambda x: x.name)
```

because here you're performing a `search` which, in theory, is just one, atomic step of `separate_dimensions`

Reply:
IIrc, what's required here is more picky than `pull_dims` (pull dims selects more dimensions than are needed). It just wants dimensions in the indices, conditional parent, and conditional condition